### DISCUSSION TITLE
Question about the output of `Hall of Frame`

### DISCUSSION THREAD

**Discussion by qizzhao** [MilesCranmer/PySR#592] [Q&A] [answered]
Hi, I am new to the SymbolicRegression. I have questions about how to control the output in hall of frame. 

Q1:  Is that possible to control the number of output in HOF?  (I tried to change the number of `topn` . It seems not working.) If yes, how? 

Q2:  It seems that each complexity has an unique result. Is that possible to give as many good results as possible , even that they have the same complexity.

Q3: Sometimes, HOF will return formula like this  `y = constant_number`, which is not I want. I want formula which has at least one variable x. How could I fix this?

Q4: I am training on a timeseries dataset and use custom loss function on Symbolic Regression Model. How could I input extra parameter (list of datetime) to my custom loss function? 

Thank you.

**Comment by MilesCranmer** [comment, marked as answer]
Q1-3 => It's a bit hard to do this because of things like the fact that $x + 1.0$ and $x + 1.000000001$ are technically different expressions with perhaps identical losses. And also things like $x / 2.0$ compared to $x * 0.5$. So the thing done right now is that only the first expression to hit a particular (loss, complexity) threshold is taken for each complexity from 1 to `maxsize`. I confess I'm not really sure how to do otherwise – would you create some way of gauging diversity of expressions stored at each complexity. 

Q4 => can you convert them to floating point numbers for this?

**Reply by tlmakinen** [reply]
been thinking about this as well -- would it be possible to implement a `minsize` or minimum complexity threshold for an expression to induct it into the hall of fame ?

**Reply by qizzhao** [reply]
Sorry for the late reply. For Q4,  yes, I could convert them to floating point number and tried to input them as `weight` parameter.  
For Q3, is there any possible way I  could avoid the case that final regression is a constant number. ( for example, y = 1.0 ). I have tried to add constraint in my custom loss function that ` if regression's prediction is all same number or the variance of prediction is equal to zero,  the loss would be inf `. But sometimes HOF will still return constant-number-only regression.

**Reply by MilesCranmer** [reply]
Just wondering — why do you need to avoid the constant expression? It does correctly express the most accurate expression at that given complexity. But why not simply only consider the `.equation_` with complexity 3 and above? All of the operations, like `predict`, `sympy`, `latex`, `latex_table` , etc, all have an argument that let you pick which equation to evaluate. So you can choose the index to avoid the simplest one.

**Reply by tlmakinen** [reply]
Yeah I've been thresholding after the search -- is there an easy way to get the HOF output file to save more equations (perhaps ~100 or so) ? i'm doing a post-search pruning with a different criterion and it'd be helpful to have more candidate expressions on hand

**Reply by MilesCranmer** [reply]
I'm not sure how to build that list of expressions though as they might be identical to the original expression (maybe you swap `a+b` to `b+a`). Maybe there's some diversity criterion for building it? 

I guess you could look at diversity in predictions outside the domain. But it seems pretty hard to write a general approach for this. Maybe the user could provide something...
